exporter/prometheusremotewriteexporter: import and add official compliance tests.

kirbyquerby opened the pull request:
**Description:** exporter/prometheusremotewrite: import and add official compliance tests.
This change imports and runs the official Prometheus Remote Write compliance tests from https://github.com/prometheus/compliance against a live collector end-to-end. **Due to data races in the official tests, this test is currently skipped.**

**Link to tracking Issue:** Fixes open-telemetry/wg-prometheus/issues/34

**Testing:** This change itself is a test :)

kirbyquerby commented:
/cc @odeke-em

odeke-em commented on exporter/prometheusremotewriteexporter/compliance/compliance_test.go:
```diff
@@ -0,0 +1,188 @@
+			t.Error(err)
+		}
+	}()
+
```
Let's add a shutdown function as well and return it or  make this a global function
```go
	shutdown := func() {
		app.Shutdown()
		// Wait until we are closed.
		<-stateChannel
		for notClosed := true; notClosed; {
			switch state := <-stateChannel; state {
			case service.Closed:
				notClosed = false
			}
		}
	}
```

kirbyquerby replied:
Added your shutdown func as a return value and called it in runTest() instead of directly invoking app.Shutdown()

odeke-em commented on exporter/prometheusremotewriteexporter/compliance/compliance_test.go:
```diff
@@ -0,0 +1,188 @@
+	app := startCollector(t, l.Addr())
+	defer app.Shutdown()
+
+	time.Sleep(10 * time.Second)
```
Why this artificial sleep? Could we perhaps reduce this as much.

odeke-em replied:
No worries, we can keep them as they are.

odeke-em commented on exporter/prometheusremotewriteexporter/compliance/compliance_test.go:
```diff
@@ -0,0 +1,188 @@
+// Copyright  OpenTelemetry Authors
```
Minor nit, please use a single space.

odeke-em reviewed (approved):
LGTM, thank you @kirbyquerby! Minor nit, please examine the 10s artificial sleep and also add a function to properly shut down lest we'll encounter race conditions with the OpenTelemetry service as I've encountered before.

kirbyquerby reviewed (commented):
Thanks for the review Emmanuel!

odeke-em commented:
Kindly requesting y'all eyes @Aneurysm9 @bogdandrutu @tigrannajaryan.

odeke-em commented:
Kindly pinging y'all, it's been 11 days and the PR is getting stale. Kindly cc-ing @dashpole

codeboten reviewed (commented):
Code looks good, please address the conflicts. It looks like lint and the tests are failing as well:

`exporter/prometheusremotewriteexporter/compliance/compliance_test.go: Expected no more than 3 groups, got 4`

dashpole reviewed (approved):
LGTM!

kirbyquerby commented:
There are currently data races in the compliance tests (and in the collector) that cause this test to fail with `-race`, so for now it's skipped. I've created https://github.com/prometheus/compliance/issues/40 for the compliance races and can work on the races in the collector after the compliance test races are fixed).

kirbyquerby commented:
The CI check failures appear to be caused elsewhere. See this clean PR which also fails checks: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5228

kirbyquerby commented:
I was mistaken :)
After running a bunch of `go mod tidy`s, everything seems to be in order (other than loadtest >:( ). No idea why #5228 is failing. Flaky tests?

odeke-em commented:
> Code looks good, please address the conflicts. It looks like lint and the tests are failing as well:
> 
> 
> 
> `exporter/prometheusremotewriteexporter/compliance/compliance_test.go: Expected no more than 3 groups, got 4`
> 
> 

@codeboten please take a look again. The Prometheus Remote Writer test code has data races they'll need to fix. On our end though we are good to go.

odeke-em commented:
Kindly cc-ing @alolita @bogdandrutu @codeboten  @Aneurysm9 to take a look before this PR gets stale again, thank you!

codeboten reviewed (approved):
i suspect another `make gotidy` is needed here to address the conflicts.

kirbyquerby commented:
Rebased + gotidied. Thanks for the review Alex!

kirbyquerby committed:
exporter/prometheusremotewriteexporter: import and add official compliance tests.


run make gotidy


run make gotidy again

????????
run go mod tidy manually in prometheusexporter


run make gotidy after rebasing


make gotidy


manually run go mod tidy in prometheusexporter

kirbyquerby committed:
make gotidy after rebase

kirbyquerby committed:
manual go mod tidy in prometheusexporter

kirbyquerby committed:
add replace directive + add context to Run() call

Aneurysm9 commented on exporter/prometheusremotewriteexporter/compliance/compliance_test.go:
```diff
@@ -0,0 +1,201 @@
+
+var (
+	logger = log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+	tests  = []func() cases.Test{
```
I'm not sure that duplicating the test harness here is the best approach.  Can we instead have an action that runs on PR that pulls the compliance test repo and runs the tests there against a build artifact?

tigrannajaryan commented:
Please resolve the conflicts and address/resolve remaining open comments.

codeboten reviewed (commented):
@kirbyquerby can this be closed as superseeded by https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5739?
